src/passes/Directize.cpp

```diff
@@ -163,11 +206,50 @@ struct Directize : public Pass {
       return;
     }
 
-    // Find which tables are valid to optimize on. They must not be imported nor
-    // exported (so the outside cannot modify them), and must have no sets in
-    // any part of the module.
+    // TODO: consider a per-table option here
+    auto initialContentsImmutable =
+      runner->options.getArgumentOrDefault(
+        "directize-initial-contents-immutable", "") != "";
+
+    // Set up the initial info.
+    TableInfoMap tables;
+    for (auto& table : module->tables) {
+      tables[table->name].initialContentsImmutable = initialContentsImmutable;
+      tables[table->name].flatTable =
+        std::make_unique<TableUtils::FlatTable>(*module, *table);
+    }
+
+    // Next, look at the imports and exports.
+
+    for (auto& table : module->tables) {
+      if (table->imported()) {
+        tables[table->name].mayBeModified = true;
+      }
+    }
+
+    for (auto& ex : module->exports) {
+      if (ex->kind == ExternalKind::Table) {
+        tables[ex->value].mayBeModified = true;
+      }
+    }
+
+    // This may already be enough information to know that we can't optimize
+    // anything. If so, skip scanning all the module contents.
+    auto canOptimize = [&]() {
+      for (auto& [_, info] : tables) {
+        if (info.canOptimize()) {
+          return true;
+        }
+      }
+      return false;
+    };
```

Comment:
Seems like this kind of sub-routine extraction could be done either with a private helper method or with a C++ closure.   Is there some rationale we can/should use for choosing one over the other? 

My understanding is that under the hood the closure does bunch of work, including creating an entire inner class but that we can/should trust the compiler to then eliminate that complexity?   What about debuggability?  Would debugging this in a debugger be any harder than a (arguably more straight forward) private helper method?

Reply:
I haven't noticed debuggability differences. I think there might be some overhead at least in older compilers here (sometimes they malloc, but not sure when). But overall in a place like this where speed doesn't matter (it's iterating on the tables, which is orders of magnitude smaller than the code) I think it's ok to use a closure, when it is convenient and shorter to write.